designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,336 @@
+
+-   New functionality
+    -   `aws-sam` Debug Configurations provide a way to launch debug sessions that articulate what is being debugged
+    -   Added CodeLenses to help users to produce Debug Configurations
+-   Removed functionality
+    -   Using CodeLenses as a mechanism to launch debug sessions will be removed. Users familiar with the outgoing functionality are provided with an assisted transition path towards using Debug Configurations.
+        -   the toolkit detects if users were using the previous CodeLens functionality based on the presence of a CodeLens configuration file (`{workspace}/.aws/templates.json`). For these users, lambda handlers that used to show Run/Debug CodeLenses will be decorated to mention the functionality shift, and will offer to convert the function's old CodeLens configuration over to a new debug configuration.
+
+### <a id="sample-debug-configurations"></a> Sample Debug Configurations
+
+Configuration structures are modelled so that they group related parameters, and allows the toolkit to reuse structure types where possible.
+
+Here is an example Debug Configuration to debug a SAM Template resource called "HelloWorldResource".
+The required fields are: type, request, invokeTarget
+
+```jsonc
+{
+    "configurations": [
+        {
+            "name": "Debug HelloWorldResource", // Users name the entry; shown in Debug dropdown
+            "type": "aws-sam",
+            // direct-invoke is the "aws-sam" variation for debugging SAM Template resources and Lambda handlers
+            "request": "direct-invoke",
+            // Reference to the thing (Template or Code) being invoked
+            "invokeTarget": {
+                "target": "template", // template | code, influences fields expected by toolkit
+                "samTemplatePath": "path to template yaml file",
+                "samTemplateResource": "HelloWorldResource" // Name of Template resource to debug
+            },
+            // Lambda Execution related arguments
+            "lambda": {
+                // Environment Variables accessible by Lambda handler
+                "environmentVariables": {
+                    "envvar1": "somevalue",
+                    "envvar2": "..."
+                },
+                // The event passed to the Lambda handler (defaults to an empty JSON object)
+                "event": {
+                    // path or json, not both
+                    "path": "somepath", // Path to event data
+                    "json": {
+                        // event data
+                    }
+                }
+            },
+            // SAM Template and SAM CLI related arguments
+            "sam": {
+                "containerBuild": false,
+                "skipNewImageCheck": false,
+                "dockerNetwork": "aaaaa",
+                "buildArguments": "--foo",
+                "localArguments": "--foo",
```

Comment:
Should these be represented as an array, an object, or a object with arrays as values?

Reply:
I prefer a single string. Curious for @abrooksv thoughts.